### PR TITLE
Compress on the fly

### LINKED ISSUE #81: Actually compress the outputs
We call our files `pb.gz` but really they are just `.pb` and we should compress them with gzip.

### PR DESCRIPTION
**Pull request by vchuravy**
Fixes #81

### CODE REVIEW

**Review by NHDaly — APPROVED** [review]
Oh, huh. TIL! :) Thanks Valentin.

File: test/PProf.jl [hunk 1]
```diff
@@ -39,7 +40,12 @@ end
     outf = pprof(out=out, web=false)
 
     # Read the exported profile
-    prof = open(io->decode(ProtoDecoder(io), PProf.perftools.profiles.Profile), outf, "r")
+    io = GzipDecompressorStream(open(outf, "r"))
+    try
+        prof = decode(ProtoDecoder(io), PProf.perftools.profiles.Profile)
+    finally
+        close(io)
+    end
```
Comment (NHDaly) [inline test/PProf.jl:48]:
```suggestion
    prof = try
        decode(ProtoDecoder(io), PProf.perftools.profiles.Profile)
    finally
        close(io)
    end
```

File: test/flamegraphs.jl [hunk 2]
```diff
@@ -42,7 +43,12 @@ end
     outf = pprof(fg, out=out, web=false)
 
     # Read the exported profile
-    fg_prof = open(io->decode(ProtoDecoder(io), PProf.perftools.profiles.Profile), outf, "r")
+    io = GzipDecompressorStream(open(outf, "r"))
+    try
+        fg_prof = decode(ProtoDecoder(io), PProf.perftools.profiles.Profile)
+    finally
+        close(io)
+    end
```
Comment (NHDaly) [inline test/flamegraphs.jl:51]:
```suggestion
    fg_prof = try
        decode(ProtoDecoder(io), PProf.perftools.profiles.Profile)
    finally
        close(io)
    end
```

File: test/Allocs.jl [hunk 3]
```diff
@@ -20,7 +21,12 @@ const out = tempname()
     outf = PProf.Allocs.pprof(out=out, web=false)
 
     # Read the exported profile
-    prof = open(io->decode(ProtoDecoder(io), PProf.perftools.profiles.Profile), outf, "r")
+    io = GzipDecompressorStream(open(outf, "r"))
+    try
+        prof = decode(ProtoDecoder(io), PProf.perftools.profiles.Profile)
+    finally
+        close(io)
+    end
```
Comment (NHDaly) [inline test/Allocs.jl:29]:
```suggestion
    prof = try
        decode(ProtoDecoder(io), PProf.perftools.profiles.Profile)
    finally
        close(io)
    end
```

**Review by NHDaly — APPROVED** [review]

**Comment by NHDaly** [comment]
🤔 hrm tests are failing now, even though we are decoding. i'm not sure why. Any idea?

**Comment by NHDaly** [comment]
@adnan-alhomssi: This seems like a good idea, and is related to your suggestion that we should be compressing the profiles and heapsnapshots before sending them back from ProfileEndpoints.jl. 

Can you have a look and see if you can help us figure out why the tests aren't passing?

**Comment by NHDaly** [comment]
Request for review: @Drvi

File: test/PProf.jl
Comment (Drvi) [inline test/PProf.jl:1]: You also need to decompress in the helper function `load_prof_proto` on line 56:

```diff
-open(io->decode(ProtoDecoder(io), PProf.perftools.profiles.Profile), file, "r")
+open(io->decode(ProtoDecoder(GzipDecompressorStream(io)), PProf.perftools.profiles.Profile), file, "r")
```

Comment (NHDaly) [reply test/PProf.jl:1]: I fixed it. Please feel free to just push a commit yourself next time! :)

Comment (NHDaly) [reply test/PProf.jl:1]: Thanks for identifying the issue 👍

**Comment by Drvi** [comment]
Will this break reading profiles from older PProf.jl versions? Typically with text files, one can sniff the gzip "magic number", but with ProtoBuf I think we need to check first if there are possible collisions with the encoded message.

**Comment by NHDaly** [comment]
Thankfully, i don't think PProf.jl ever reads profiles, we just hand them over to `pprof`, which seems to be able to read them fine, in either version.

So i think it's okay? What concern did you have in mind?

**Comment by Drvi** [comment]
Ah if we don't need to read it on Julia side, then it's fine:+1: sorry for the noise

**Pull request merged**